pkg/applicationserver/io/pubsub/pubsub.go

```diff
@@ -235,27 +237,31 @@ func (ps *PubSub) start(ctx context.Context, pb *ttnpb.ApplicationPubSub) (err e
 	defer func() {
 		cancel(err)
 	}()
+
 	i := &integration{
 		ApplicationPubSub: *pb,
 		ctx:               ctx,
 		cancel:            cancel,
 		closed:            make(chan struct{}),
 		server:            ps.server,
 	}
+	defer close(i.closed)
 	if _, loaded := ps.integrations.LoadOrStore(psUID, i); loaded {
-		log.FromContext(ctx).Debug("Integration already started")
+		logger.Debug("Pub/sub already started")
 		return errAlreadyConfigured.WithAttributes("application_uid", appUID, "pub_sub_id", pb.PubSubID)
 	}
-	go func() {
-		<-ctx.Done()
```

Comment:
Deleting the `psUID` from `integrations` here allows a new integration to start even though the old one has not been stopped/disconnected completely yet. So instead of using a separate goroutine that basically races with the default shutdown procedure, we allow the task itself to do the cleanup (in the refactored version).